R/selector.R

```diff
@@ -0,0 +1,279 @@
+    if (!is.null(element)) {
+      selector <- str_remove(selector, elementRegex)
+    }
+
+    ## https://www.w3.org/TR/CSS21/syndata.html#value-def-identifier
+    ##  In CSS, identifiers (including element names, classes, and IDs in selectors) can contain only the characters [a-zA-Z0-9] and ISO 10646 characters U+00A0 and higher, plus the hyphen (-) and the underscore (_); they cannot start with a digit, two hyphens, or a hyphen followed by a digit. Identifiers can also contain escaped characters and any ISO 10646 character as a numeric code (see next item). For instance, the identifier "B&W?" may be written as "B\&W\?" or "B\26 W\3F".
+    # # define simpler (maybe not accurate) regex
+    # id_regex <- "^#[^#.:[\\s]+" # `#` then everything that isn't a `#`, `.`, `:`, or white space
+    # class_regex <- "^\\.[^#.:[\\s]+" # `.` then everything that isn't a `.`, `:`, or white space
+
+    tmpId <- str_match_first(selector, "#[^.:[]+")
+    if (!is.null(tmpId)) {
+      id <- str_remove(tmpId, "^#")
+      selector <- str_remove(selector, tmpId, fixed = TRUE)
+    }
+
+    classes <- str_remove(str_match_all(selector, "\\.[^.:[]+"), "^\\.")
+    if (length(classes) == 0) {
+      classes <- NULL
+    }
+    # if (!is.null(classes)) {
+    #   selector <- str_remove(selector, "\\.[^.:[]+")
+    # }
+  }
+
+  structure(class = selectorClass, list(
+    element = element,
+    id = id,
+    classes = classes,
+    type = type,
+    traversal = traversal
+  ))
+}
+
+
+asSelectorList <- function(selector) {
+  if (isSelectorList(selector)) {
+    return(selector)
+  }
+  if (is.character(selector)) {
+    selector <- asSelector(selector)
+  }
+  if (isSelector(selector)) {
+    selector <- list(selector)
+  }
+  if (!is.list(selector)) {
+    stop("Do not know how to convert non list object into a `htmltools.selector.list`")
+  }
+
+  isSelectorVals <- vapply(selector, isSelector, logical(1))
+  if (!all(isSelectorVals)) {
+    stop("Can only convert a list of selectors to a `htmltools.selector.list`")
+  }
+  structure(class = selectorListClass, selector)
+}
+
+#' @export
+format.htmltools.selector <- function(x, ...) {
+  paste0(
+    c(
+      if (x$traversal == SELECTOR_CHILD) "> ",
+      if (x$type == SELECTOR_EVERYTHING) {
+        "*"
+      } else {
+        paste0(c(
+          x$element,
+          if (!is.null(x$id)) paste0("#", x$id),
+          if (!is.null(x$classes)) paste0(".", x$classes)
+        ))
+      }
+    ),
+    collapse = ""
+  )
+}
+#' @export
+format.htmltools.selector.list <- function(x, ...) {
+  paste0(unlist(lapply(x, format, ...)), collapse = " ")
+}
+
+#' @export
+print.htmltools.selector <- function(x, ...) {
+  cat("// htmltools css selector\n")
+  cat(format(x, ...), "\n")
+}
+#' @export
+print.htmltools.selector.list <- function(x, ...) {
+  cat("// htmltools css selector list\n")
+  cat(format(x, ...), "\n")
+}
+
+
+
+str_replace <- function(text, pattern, replacement, fixed = FALSE) {
```

Comment:
I don't think it makes sense to add the `fixed` parameter, since it doesn't exist in `stringr::str_replace()` -- if you want to use a parameter that's in `sub()`, why not just use `sub()`?

Same for `str_replace_all` below.